File: backend/app/schemas/document.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, field_validator

from app.models.enums import AuthorityLevel, CollectionType, ConfidentialityLevel, EmbeddingStatus, ParseStatus


def _require_http_scheme(url: str | None) -> str | None:
    # Report exports turn this into a real <a href>/Markdown link, so a
    # non-http(s) scheme here (e.g. "javascript:...") would be a stored-XSS
    # vector the moment someone views the report -- reject at the boundary
    # where a human enters it, not just at render time.
    if url and not (url.startswith("http://") or url.startswith("https://")):
        raise ValueError("Source URL must start with http:// or https://")
    return url


class SourceDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID | None
    collection_type: CollectionType
    source_type: str | None
    authority_level: AuthorityLevel | None
    title: str
    issuer: str | None
    url: str | None
    original_filename: str | None
    mime_type: str | None
    jurisdiction: str | None
    document_category: str | None
    publication_date: date | None
    effective_date: date | None
    expiration_date: date | None
    version: str | None
    is_current: bool
    is_superseded: bool
    sha256: str | None
    parse_status: ParseStatus
    embedding_status: EmbeddingStatus
    confidentiality_level: ConfidentialityLevel
    parse_error: str | None
    created_at: datetime
    updated_at: datetime


class SourceDocumentMetadataUpdate(BaseModel):
    title: str | None = None
    issuer: str | None = None
    url: str | None = None
    jurisdiction: str | None = None
    document_category: str | None = None
    source_type: str | None = None
    authority_level: AuthorityLevel | None = None
    publication_date: date | None = None
    effective_date: date | None = None
    expiration_date: date | None = None
    version: str | None = None
    is_current: bool | None = None
    confidentiality_level: ConfidentialityLevel | None = None

    @field_validator("url")
    @classmethod
    def _validate_url(cls, value: str | None) -> str | None:
        return _require_http_scheme(value)


class SourceDocumentWithProject(SourceDocumentRead):
    project_name: str | None


class SourceChunkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    chunk_index: int
    text: str
    token_count: int | None
    page_number: int | None
    section_title: str | None
    heading_path: str | None
    citation_label: str

File: frontend/src/types/document.ts
export type CollectionType = "COMPANY" | "AUTHORITY" | "THIRD_PARTY" | "COMPETITOR";
export type ParseStatus = "PENDING" | "PROCESSING" | "COMPLETE" | "FAILED" | "QUARANTINED";
export type EmbeddingStatus = "PENDING" | "PROCESSING" | "COMPLETE" | "FAILED" | "STALE";
export type ConfidentialityLevel = "PUBLIC" | "INTERNAL" | "RESTRICTED";

export interface SourceDocument {
  id: string;
  project_id: string | null;
  collection_type: CollectionType;
  source_type: string | null;
  authority_level: string | null;
  title: string;
  issuer: string | null;
  url: string | null;
  original_filename: string | null;
  mime_type: string | null;
  jurisdiction: string | null;
  document_category: string | null;
  publication_date: string | null;
  effective_date: string | null;
  expiration_date: string | null;
  version: string | null;
  is_current: boolean;
  is_superseded: boolean;
  sha256: string | null;
  parse_status: ParseStatus;
  embedding_status: EmbeddingStatus;
  confidentiality_level: ConfidentialityLevel;
  parse_error: string | null;
  created_at: string;
  updated_at: string;
}

export interface SourceDocumentWithProject extends SourceDocument {
  project_name: string | null;
}

export interface SourceChunk {
  id: string;
  document_id: string;
  chunk_index: number;
  text: string;
  token_count: number | null;
  page_number: number | null;
  section_title: string | null;
  heading_path: string | null;
  citation_label: string;
}

export type JobStatus = "QUEUED" | "RUNNING" | "COMPLETE" | "FAILED" | "CANCELLED";

export interface Job {
  id: string;
  job_type: string;
  project_id: string | null;
  related_id: string | null;
  status: JobStatus;
  progress_percent: number;
  current_stage: string | null;
  logs: Array<{ stage?: string; message: string }>;
  error_summary: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}
